File: Data Structures & Algorithms/max-area-of-island/submission-1.py
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        dir=[[1,0],[0,1],[-1,0],[0,-1]]

        row=len(grid)
        col=len(grid[0])
        maxAns=0
        
        def dfs(r,c):
            if(r<0 or r>=row or c<0 or c>=col or grid[r][c]==0):
                return 0

            grid[r][c]=0
            ans=1
            for dr,dc in dir:
                ans=ans+dfs(r+dr,c+dc)
            return ans

        for r in range(row):
            for c in range(col):
                if grid[r][c]==1:
                    maxAns = max(maxAns,dfs(r,c))

        return maxAns

File: Data Structures & Algorithms/rotting-fruit/submission-1.cpp
class Solution {
public:
    int dir[4][2] = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
    int orangesRotting(vector<vector<int>>& grid) {
        int n = grid.size();
        int m = grid[0].size();
        queue<pair<int,int>> q;
        int numFresh=0;
        for(int i=0;i<n;i++){
            for(int j=0;j<m;j++){
                if(grid[i][j]==2)q.push({i,j});
                if(grid[i][j]==1)numFresh++;
            }
        }
        int ans=0;
        

        while(numFresh> 0 && !q.empty()){
            int l = q.size();
            for(int i=0;i<l;i++){
            int row = q.front().first;
            int col = q.front().second;
            q.pop();
            for(int i=0;i<4;i++){
                int r = row+dir[i][0];
                int c = col+dir[i][1];
                if(r<0 || c<0 || r==n || c==m || grid[r][c]!=1)continue;
                grid[r][c]=2;
                numFresh--;
                q.push({r,c});
            }}
            ans++;
        }
        if(numFresh)return -1;
        return ans;
    }
};
